# make_bins

These instructions are based on https://github.com/susbo/Drosophila_unistrand_clusters/tree/main/De-novo_clusters

Analysis to quantify genome-wide LTR/Gypsy content. Please note that [hgLoadOut](https://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/hgLoadOut) and bedtools has to be available in your PATH.

## Analysis steps

To produce the output files, run the following scripts:

```
. 01_fix_headers.sh
```
*Step 1:* Restore the full contig names to `EDTA/genome.fa.mod.EDTA.anno/genome.fa.mod.out`, since EDTA will cap them at a fixed number of characters.

```
. 02_tab.sh
```
*Step 2:* Convert the .out file generated by RepeatMasker (run by EDTA) to a tab-separated format.

```
. 03_bed.sh
```
*Step 3:* Convert the tab-separated format to bed format, only including LTR/Gypsy transposons, and use bedtools merge to combine overlapping bed regions. This step uses `toBed6-10.pl` which can be retrieved from [http://genomewiki.ucsc.edu/index.php/RepeatMasker](http://genomewiki.ucsc.edu/index.php/RepeatMasker).

```
. 04_bins.sh
```
*Step 4:* Calculate Gypsy coverage per 2.5 kb, 5 kb, and 10 kb window. The results are found in `out/04_bins/overlap/bins_2.5k`, `out/04_bins/overlap/bins_5k`, and `out/04_bins/overlap/bins_10k`, which can be copied to the `bins` directory.


## Output

The output is produced in the form of a tab-separated file representing
1. Chromosome
2. Start coordinate
3. End coordinate
4. Number of overlapping bed intervals
5. Length of bin (usually 100000)
6. Transposon coverage (within 0..1)

Example output:
```
$ head out/04_bins/overlap/bins_5k/plus/Dmel.dm6.bed 
chr2L   0       5000    0       0       5000    0.0000000
chr2L   5000    10000   0       0       5000    0.0000000
chr2L   10000   15000   0       0       5000    0.0000000
chr2L   15000   20000   0       0       5000    0.0000000
chr2L   20000   25000   0       0       5000    0.0000000
chr2L   25000   30000   0       0       5000    0.0000000
chr2L   30000   35000   0       0       5000    0.0000000
chr2L   35000   40000   0       0       5000    0.0000000
chr2L   40000   45000   0       0       5000    0.0000000
chr2L   45000   50000   0       0       5000    0.0000000
$ head out/04_bins/overlap/bins_5k/minus/Dmel.dm6.bed 
chr2L   0       5000    0       0       5000    0.0000000
chr2L   5000    10000   0       0       5000    0.0000000
chr2L   10000   15000   0       0       5000    0.0000000
chr2L   15000   20000   0       0       5000    0.0000000
chr2L   20000   25000   0       0       5000    0.0000000
chr2L   25000   30000   0       0       5000    0.0000000
chr2L   30000   35000   0       0       5000    0.0000000
chr2L   35000   40000   0       0       5000    0.0000000
chr2L   40000   45000   0       0       5000    0.0000000
chr2L   45000   50000   0       0       5000    0.0000000
```

## Example

EDTA annotations for Drosophila melanogaster has been provided as an example. The genome assmebly was too large to provide, but can be downloaded using

```
wget https://hgdownload.soe.ucsc.edu/goldenPath/dm6/bigZips/dm6.fa.gz -O species/Dmel/dm6/genome.fa.gz
gunzip species/Dmel/dm6/genome.fa.gz
samtools faidx species/Dmel/dm6/genome.fa
```

Following the instructions above should result in the following files:

```
$ tree out/
out/
├── 01_fix_header
│   └── Dmel.dm6.out
├── 02_tab
│   ├── all
│   │   └── Dmel.dm6.tab
│   ├── minus
│   │   └── Dmel.dm6.tab
│   └── plus
│       └── Dmel.dm6.tab
├── 03_bed
│   ├── minus
│   │   ├── Dmel.dm6.bed
│   │   └── merged
│   │       └── Dmel.dm6.bed
│   └── plus
│       ├── Dmel.dm6.bed
│       └── merged
│           └── Dmel.dm6.bed
└── 04_bins
    ├── bins
    │   ├── Dmel.dm6.10000.bed.gz
    │   ├── Dmel.dm6.2500.bed.gz
    │   └── Dmel.dm6.5000.bed.gz
    └── overlap
        ├── bins_10k
        │   ├── minus
        │   │   └── Dmel.dm6.bed
        │   └── plus
        │       └── Dmel.dm6.bed
        ├── bins_2.5k
        │   ├── minus
        │   │   └── Dmel.dm6.bed
        │   └── plus
        │       └── Dmel.dm6.bed
        └── bins_5k
            ├── minus
            │   └── Dmel.dm6.bed
            └── plus
                └── Dmel.dm6.bed

22 directories, 17 files
```
