# make_bins

These instructions are based on https://github.com/susbo/Drosophila_unistrand_clusters/tree/main/De-novo_clusters and describe how to construct BED-like files with Gypsy/LTR content to be used as input for FlaHMM.

**Input:** A RepeatMasker output file (.out) for the genome of interest.<br>
**Output:** Two BED-like files (plus and minus strand) with LTR/Gypsy content per genomic bin.<br>

*Please note that [hgLoadOut](https://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/hgLoadOut) and bedtools has to be available in your PATH.*

## Analysis steps

The pipeline below describes how to produce the BED-like files from EDTA output, but can be easily modified (as described) to use any RepeatMasker output files.

```
. 01_fix_headers.sh
```
*Step 1:* Restore the full contig names to `EDTA/genome.fa.mod.EDTA.anno/genome.fa.mod.out`, since EDTA will cap them at a fixed number of characters. Please note that this step is only relevant if using EDTA de-novo annotation output and assumes that files named `species/$species/$build/annotation/EDTA/genome.fa.mod.EDTA.anno/genome.fa.mod.out` are available.

If using RepeatMasker output from another source than EDTA, simply rename and copy your `.out` files into `species/$species/$build/annotation/EDTA/genome.fa.mod.EDTA.anno/genome.fa.mod.out` where `$species` may refer to the species and `$build` may refer to the genome assembly. However, any strings are valid as "$species/$build", e.g., "Dmel/dm6" or "My_secret_species/01".

```
. 02_tab.sh
```
*Step 2:* Convert the .out file generated by RepeatMasker into a tab-separated format.

```
. 03_bed.sh
```
*Step 3:* Convert the tab-separated format to bed format, only including LTR/Gypsy transposons, and use bedtools merge to combine overlapping bed regions. This step uses `toBed6-10.pl` which can be retrieved from [http://genomewiki.ucsc.edu/index.php/RepeatMasker](http://genomewiki.ucsc.edu/index.php/RepeatMasker).

```
. 04_bins.sh
```
*Step 4:* Calculate Gypsy coverage per 2.5 kb, 5 kb, and 10 kb window. The results are found in `out/04_bins/overlap/bins_2.5k`, `out/04_bins/overlap/bins_5k`, and `out/04_bins/overlap/bins_10k`, which can be copied to the `bins` directory.


## Output file format

The output is produced in the form of a tab-separated file representing
1. Chromosome
2. Start coordinate
3. End coordinate
4. Number of overlapping bed intervals
5. Length of bin
6. Transposon coverage (within 0..1)

Example output:
```
$ head out/04_bins/overlap/bins_5k/plus/Dmel.dm6.bed 
chr2L   0       5000    0       0       5000    0.0000000
chr2L   5000    10000   0       0       5000    0.0000000
chr2L   10000   15000   0       0       5000    0.0000000
chr2L   15000   20000   0       0       5000    0.0000000
chr2L   20000   25000   0       0       5000    0.0000000
chr2L   25000   30000   0       0       5000    0.0000000
chr2L   30000   35000   0       0       5000    0.0000000
chr2L   35000   40000   0       0       5000    0.0000000
chr2L   40000   45000   0       0       5000    0.0000000
chr2L   45000   50000   0       0       5000    0.0000000
$ head out/04_bins/overlap/bins_5k/minus/Dmel.dm6.bed 
chr2L   0       5000    0       0       5000    0.0000000
chr2L   5000    10000   0       0       5000    0.0000000
chr2L   10000   15000   0       0       5000    0.0000000
chr2L   15000   20000   0       0       5000    0.0000000
chr2L   20000   25000   0       0       5000    0.0000000
chr2L   25000   30000   0       0       5000    0.0000000
chr2L   30000   35000   0       0       5000    0.0000000
chr2L   35000   40000   0       0       5000    0.0000000
chr2L   40000   45000   0       0       5000    0.0000000
chr2L   45000   50000   0       0       5000    0.0000000
```

## Example

#### Using EDTA output files

EDTA annotations for Drosophila melanogaster has been provided as an example. The genome assembly was too large to provide, but can be downloaded using

```
wget https://hgdownload.soe.ucsc.edu/goldenPath/dm6/bigZips/dm6.fa.gz -O species/Dmel/dm6/genome.fa.gz
gunzip species/Dmel/dm6/genome.fa.gz
samtools faidx species/Dmel/dm6/genome.fa
```

#### Using USCS pre-computed RepeatMasker output files

Aside from EDTA, any RepeatMasker output files can be used as input. Below is an example of an .out file downloaded from the UCSC Genome Browser.

```
# Download pre-calculated RepeatMasker output for dm6
wget https://hgdownload.soe.ucsc.edu/goldenPath/dm6/bigZips/dm6.fa.out.gz
gunzip dm6.fa.out.gz

# Create folders and add RepeatMasker output file to the expected folder. Please 
# note that we chosen "Dmel" as the species name and "dm6-rm" as the build name
# but this could be any strings.
mkdir species/Dmel/dm6-rm
mkdir species/Dmel/dm6-rm/annotation
mkdir species/Dmel/dm6-rm/annotation/EDTA
mkdir species/Dmel/dm6-rm/annotation/EDTA/genome.fa.mod.EDTA.anno
mv dm6.fa.out species/Dmel/dm6-rm/annotation/EDTA/genome.fa.mod.EDTA.anno/genome.fa.mod.out
```

Moreover, you need to provide the genome.fa and genome.fa.fai for the genome of interest. If downloading them using the instructions above, they can be copied into the new folder using
```
# Copy genome from Dmel/dm6 into Dmel/dm6-rm
cp species/Dmel/dm6/genome.fa* species/Dmel/dm6-rm/
```

#### Expected input file structure

If preparing both the EDTA and UCSC-RepeatMasker examples, the following directory structure would be expected in the `species` directory

```

$ tree species
species/
└── Dmel
    ├── dm6
    │   ├── annotation
    │   │   └── EDTA
    │   │       └── genome.fa.mod.EDTA.anno
    │   │           └── genome.fa.mod.out
    │   ├── genome.fa
    │   └── genome.fa.fai
    └── dm6-rm
        ├── annotation
        │   └── EDTA
        │       └── genome.fa.mod.EDTA.anno
        │           └── genome.fa.mod.out
        ├── genome.fa
        └── genome.fa.fai

9 directories, 6 files
```

#### Expected results

Following the instructions above (step 01 to 04) should result in the following files

```
out/
├── 01_fix_header
│   ├── Dmel.dm6.out
│   └── Dmel.dm6-rm.out
├── 02_tab
│   ├── all
│   │   ├── Dmel.dm6-rm.tab
│   │   └── Dmel.dm6.tab
│   ├── minus
│   │   ├── Dmel.dm6-rm.tab
│   │   └── Dmel.dm6.tab
│   └── plus
│       ├── Dmel.dm6-rm.tab
│       └── Dmel.dm6.tab
├── 03_bed
│   ├── minus
│   │   ├── Dmel.dm6.bed
│   │   ├── Dmel.dm6-rm.bed
│   │   └── merged
│   │       ├── Dmel.dm6.bed
│   │       └── Dmel.dm6-rm.bed
│   └── plus
│       ├── Dmel.dm6.bed
│       ├── Dmel.dm6-rm.bed
│       └── merged
│           ├── Dmel.dm6.bed
│           └── Dmel.dm6-rm.bed
└── 04_bins
    ├── bins
    │   ├── Dmel.dm6.10000.bed.gz
    │   ├── Dmel.dm6.2500.bed.gz
    │   ├── Dmel.dm6.5000.bed.gz
    │   ├── Dmel.dm6-rm.10000.bed.gz
    │   ├── Dmel.dm6-rm.2500.bed.gz
    │   └── Dmel.dm6-rm.5000.bed.gz
    └── overlap
        ├── bins_10k
        │   ├── minus
        │   │   ├── Dmel.dm6.bed
        │   │   └── Dmel.dm6-rm.bed
        │   └── plus
        │       ├── Dmel.dm6.bed
        │       └── Dmel.dm6-rm.bed
        ├── bins_2.5k
        │   ├── minus
        │   │   ├── Dmel.dm6.bed
        │   │   └── Dmel.dm6-rm.bed
        │   └── plus
        │       ├── Dmel.dm6.bed
        │       └── Dmel.dm6-rm.bed
        └── bins_5k
            ├── minus
            │   ├── Dmel.dm6.bed
            │   └── Dmel.dm6-rm.bed
            └── plus
                ├── Dmel.dm6.bed
                └── Dmel.dm6-rm.bed

22 directories, 34 files
```

The content of the `04_bins/overlap` folder can be directory copied into `../../bins` and act as input files.
